plugin-server/src/main/ingestion-queues/analytics-events-ingestion-consumer.ts

```diff
@@ -78,50 +78,67 @@ export async function eachBatchIngestionWithOverflow(
         const batches = []
         const seenIds: Set<string> = new Set()
         let currentBatch: KafkaMessage[] = []
+
         for (const message of kafkaMessages) {
             const pluginEvent = formPipelineEvent(message)
             const seenKey = `${pluginEvent.team_id}:${pluginEvent.distinct_id}`
+
+            // Events with a null key should have been produced to the the
+            // KAFKA_EVENTS_PLUGIN_INGESTION_OVERFLOWtopic, so we shouldn't see them here as this consumer's
+            // topic is set to KAFKA_EVENTS_PLUGIN_INGESTION. However, there could be some lingering events
+            // from before the new *_OVERFLOW topic was initialized. Any events with a null key or that
+            // exceed capacity are redirected to the *_OVERFLOW topic.
+            if (message.key == null || ConfiguredLimiter.consume(seenKey, 1) === false) {
+                // Set message key to be null so we know to send it to overflow topic.
+                // We don't want to do it here to preserve the kafka offset handling
+                message.key = null
+                currentBatch.push(message)
+
+                // Warnings are limited to 1/key/hour to avoid spamming.
+                if (pluginEvent.team_id && WarningLimiter.consume(seenKey, 1)) {
+                    captureIngestionWarning(
+                        queue.pluginsServer.hub.db,
+                        pluginEvent.team_id,
+                        'ingestion_capacity_overflow',
+                        {
+                            overflowDistinctId: pluginEvent.distinct_id,
+                        }
+                    )
+                }
```

Comment:
nit: would be more logical for me for this to be in the eachMessage part

Reply:
We would need to parse the event and construct the `seenKey` again, and we are already doing it here.

Reply:
I am agreeing with you though... Let me push a change just to see how this looks.

Reply:
Mostly because I'm hesitant about this firing off a query during batching...